### PR TITLE
Expose memory-related options in `Config`

### LINKED ISSUE #1501: Setting vmemoryuse to 3G causes OOM errors
- What are the steps to reproduce the issue?

```
ollisg@muse:~/test/wasmtime/wasmtime-v0.15.0-x86_64-linux$ ulimit -v 3145728
ollisg@muse:~/test/wasmtime/wasmtime-v0.15.0-x86_64-linux$ cat hello.rs 
fn main() {
  println!("Hello, world!");
}
ollisg@muse:~/test/wasmtime/wasmtime-v0.15.0-x86_64-linux$ rustc hello.rs --target wasm32-wasi
ollisg@muse:~/test/wasmtime/wasmtime-v0.15.0-x86_64-linux$ ./wasmtime hello.wasm 
Error: failed to run main module `hello.wasm`

Caused by:
    0: failed to instantiate "hello.wasm"
    1: Insufficient resources: Cannot allocate memory (os error 12)
```

- What do you expect to happen? What does actually happen? Does it panic, and
  if so, with which assertion?

I don't expect such a high virtual memory limit to cause an OOM on such a small program. 

- Which Wasmtime version / commit hash / branch are you using?

This is the using the `0.15.0` tarball, but I have been able to reproduce it also using the `0.15.0` c-api via Perl/FFI, see https://github.com/perlwasm/Wasm/issues/22

- If relevant, can you include some extra information about your environment?
  (Rust version, operating system, architecture...)

I was able to reproduce on my Debian Linux x864_64 box:

```
ollisg@muse:~/test/wasmtime/wasmtime-v0.15.0-x86_64-linux$ rustc --version
rustc 1.42.0 (b8cedc004 2020-03-09)
ollisg@muse:~/test/wasmtime/wasmtime-v0.15.0-x86_64-linux$ cat /etc/debian_version 
9.12
```

cpantesters use a variety of platforms, with the binary wasmtime tarball, so I am not sure that this matters, but including for completeness.

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit was initially motivated by looking more into #1501, but it
ended up balooning a bit after finding a few issues. The high-level
items in this commit are:

* New configuration options via `wasmtime::Config` are exposed to
  configure the tunable limits of how memories are allocated and such.
* The `MemoryCreator` trait has been updated to accurately reflect the
  required allocation characteristics that JIT code expects.
* A bug has been fixed in the cranelift wasm code generation where if no
  guard page was present bounds checks weren't accurately performed.

The new `Config` methods allow tuning the memory allocation
characteristics of wasmtime. Currently 64-bit platforms will reserve 6GB
chunks of memory for each linear memory, but by tweaking various config
options you can change how this is allocate, perhaps at the cost of
slower JIT code since it needs more bounds checks. The methods are
intended to be pretty thoroughly documented as to the effect they have
on the JIT code and what values you may wish to select. These new
methods have been added to the spectest fuzzer to ensure that various
configuration values for these methods don't affect correctness.

The `MemoryCreator` trait previously only allocated memories with a
`MemoryType`, but this didn't actually reflect the guarantees that JIT
code expected. JIT code is generated with an assumption about the
minimum size of the guard region, as well as whether memory is static or
dynamic (whether the base pointer can be relocated). These properties
must be upheld by custom allocation engines for JIT code to perform
correctly, so extra parameters have been added to
`MemoryCreator::new_memory` to reflect this.

Finally the fuzzing with `Config` turned up an issue where if no guard
pages present the wasm code wouldn't correctly bounds-check memory
accesses. The issue here was that with a guard page we only need to
bounds-check the first byte of access, but without a guard page we need
to bounds-check the last byte of access. This meant that the code
generation needed to account for the size of the memory operation
(load/store) and use this as the offset-to-check in the no-guard-page
scenario. I've attempted to make the various comments in cranelift a bit
more exhaustive too to hopefully make it a bit clearer for future
readers!

Closes #1501

### CODE REVIEW

File: cranelift/wasm/src/code_translator.rs [hunk 1]
```diff
@@ -1703,25 +1705,66 @@ fn get_heap_addr(
...
+    //
+    // If we don't have a guard page of unmapped memory, though, then we can't
+    // rely on this trapping behavior through segfaults. Instead we need to
+    // bounds-check the entire memory access here which is everything from
+    // `addr32 + offset` to `addr32 + offset + width` (not inclusive). In this
+    // scenario our adjusted offset that we're checking is `offset + width`.
+    //
+    // If we have a guard page, however, then we can perform a further
+    // optimization of the generated code by only checking multiples of the
+    // offset-guard size to be more CSE-friendly. Knowing that we have a least
```
Comment (sunfishcode) [inline cranelift/wasm/src/code_translator.rs:1727]: typo: a -> at

File: cranelift/wasm/src/code_translator.rs [hunk 2]
```diff
@@ -1703,25 +1705,66 @@ fn get_heap_addr(
...
+    // 1 page of a guard page we're then able to disregard the `width` of the
+    // access because we assume bytes are accessed first-to-last and if we
+    // happen to straddle a page boundary at least one byte will trap.
```
Comment (sunfishcode) [inline cranelift/wasm/src/code_translator.rs:1730]: This first-to-last property is interesting, as optimizers could be tempted to optimize memory accesses in ways that could change the access order -- for example optimizing `load.i32(x) & 0xff000000` into a one-byte load from the last byte. Could you add some text to https://github.com/bytecodealliance/wasmtime/blob/master/cranelift/docs/ir.md#memory-operation-flags that says that when an access is not `notrap`, the bytes are defined to be accessed first-to-last, so that we at least document it?

File: crates/api/src/externals.rs [hunk 3]
```diff
@@ -887,8 +887,58 @@ pub unsafe trait LinearMemory {
 /// Note that this is a relatively new and experimental feature and it is recommended
 /// to be familiar with wasmtime runtime code to use it.
 pub unsafe trait MemoryCreator: Send + Sync {
-    /// Create new LinearMemory
-    fn new_memory(&self, ty: MemoryType) -> Result<Box<dyn LinearMemory>, String>;
+    /// Create new LinearMemory object from the specified parameters.
```
Comment (sunfishcode) [inline crates/api/src/externals.rs:890]: Code-quote `LinearMemory`.

File: crates/api/src/externals.rs [hunk 4]
```diff
@@ -887,8 +887,58 @@ pub unsafe trait LinearMemory {
...
+    /// both the minimum and maximum size, in wasm pages.
+    ///
+    /// The `reserved_size` value indicates the expected size of the
+    /// reservation that is to be made for this memory. If this value is `None`
+    /// than the implementation is free to allocate memory as it sees fit. If
+    /// the value is `Some`, however, then the implementation is expected to
+    /// reserve that many bytes for the memory's allocation, plus the guard
+    /// size at the end. In this case `grow` should never move the base pointer
+    /// and the maximum size of `ty` is guaranteed to fit within
+    /// `reserved_size`.
```
Comment (sunfishcode) [inline crates/api/src/externals.rs:902]: One subtlety that would be good to clarify here is that this is talking about reserving address space, not necessarily actual memory.

File: crates/api/src/runtime.rs [hunk 5]
```diff
@@ -385,6 +386,182 @@ impl Config {
         self.memory_creator = Some(MemoryCreatorProxy { mem_creator });
         self
     }
+
+    /// Configures the maximum size where a linear memory is considered static,
```
Comment (sunfishcode) [inline crates/api/src/runtime.rs:390]: Because wasm sometimes measures sizes in pages, we should clarify that this is a size in bytes.

File: crates/api/src/runtime.rs [hunk 6]
```diff
@@ -385,6 +386,182 @@ impl Config {
...
+        self.tunables.static_memory_offset_guard_size =
+            cmp::max(guard_size, self.tunables.static_memory_offset_guard_size);
+        self
+    }
+}
+
+fn round_up_to_pages(val: u64) -> u64 {
+    let page_size = region::page::size() as u64;
+    val.checked_add(page_size - 1)
+        .map(|val| val / page_size * page_size)
```
Comment (sunfishcode) [inline crates/api/src/runtime.rs:563]: The page size will always be a power of 2, so we can do `val & -page_size`.

File: crates/api/src/runtime.rs [hunk 7]
```diff
@@ -385,6 +386,182 @@ impl Config {
...
+        .unwrap_or(u64::max_value() / page_size + 1)
```
Comment (sunfishcode) [inline crates/api/src/runtime.rs:564]: This could produce a value which is *lower* than the given `val`, but I guess that's ok here because it'd be an unusably large value in either case.

File: src/lib.rs [hunk 8]
```diff
@@ -152,6 +152,19 @@ struct CommonOptions {
         default_value = "2",
     )]
     opt_level: wasmtime::OptLevel,
+
+    /// Maximum size of wasm memory before it becomes dynamically relocatable
+    /// instead of up-front-reserved.
```
Comment (sunfishcode) [inline src/lib.rs:157]: "size in bytes", and so on below.

File: cranelift/wasm/src/code_translator.rs (hunk 2 above)
Comment (alexcrichton) [reply cranelift/wasm/src/code_translator.rs:1730]: So thinking more about this, I think I just carried this comment over from somewhere that I didn't fully understand. I think this property isn't actually necessary (and may not be necessary to document).

The main assumption I think we can make is that all our accesses are smaller than the size of a page (e.g. the minimum size of the guard region). In that case the first byte is either:

* in wasm memory which is normal-bounds-checked and all remaining bytes are either in wasm memory or the guard-page
* in the guard page, in which case the bounds check will fail.

Does that sound right to you as well? If so I'll reword this to and probably throw in an assert that the `width` is less than 1k or something like that with a doc comment explaining.

File: crates/api/src/runtime.rs (hunk 7 above)
Comment (alexcrichton) [reply crates/api/src/runtime.rs:565]: Ah true yeah, but I was mostly just striving to do something reasonable-ish here.

File: tests/all/memory_creator.rs [hunk 9]
```diff
@@ -108,7 +108,14 @@ mod not_for_windows {
...
 
     unsafe impl MemoryCreator for CustomMemoryCreator {
-        fn new_memory(&self, ty: MemoryType) -> Result<Box<dyn LinearMemory>, String> {
+        fn new_memory(
+            &self,
+            ty: MemoryType,
+            reserved_size: Option<u64>,
+            guard_size: u64,
+        ) -> Result<Box<dyn LinearMemory>, String> {
+            assert_eq!(guard_size, 0);
```
Comment (eust-dfinity) [inline tests/all/memory_creator.rs:117]: Actually that example is allocating one guard page

File: cranelift/wasm/src/code_translator.rs (hunk 2 above)
Comment (alexcrichton) [reply cranelift/wasm/src/code_translator.rs]: I went ahead and tried to update the comment with my thinking here.

File: tests/all/memory_creator.rs (hunk 9 above)
Comment (alexcrichton) [reply tests/all/memory_creator.rs:117]: True yeah, I was mostly just getting this test to pass. Since the JIT is configured that there might not be a guard page then the guard page allocation here shouldn't be necessary.

**Comment by sunfishcode** [comment]
Looks good!

File: crates/c-api/src/config.rs [hunk 10]
```diff
@@ -146,3 +146,18 @@ pub unsafe extern "C" fn wasmtime_config_cache_config_load(
...
     )
 }
+
+#[no_mangle]
+pub extern "C" fn wasmtime_config_static_memory_maximum_size_set(c: &mut wasm_config_t, size: u64) {
+    c.config.static_memory_maximum_size(size);
+}
+
+#[no_mangle]
+pub extern "C" fn wasmtime_config_static_memory_guard_size(c: &mut wasm_config_t, size: u64) {
```
Comment (plicease) [inline crates/c-api/src/config.rs:156]: After testing this from Perl, I think the `_set` suffix was accidentally omitted from this declaration.  Same for `wasmtime_config_dynamic_memory_guard_size` below.
